### ISSUE TITLE
MQTT  - based scaler

### ISSUE THREAD

**Issue by arschles** [kedacore/keda#1282]
I'd like to have a scaler listen to an MQTT source and scale up/down based on incoming messages. This would be useful for Kubernetes clusters that run on the edge (for example, in IoT scenarios and in scenarios where k8s is running on underpowered machines)

  - **Scaler Source:** A scaler to support scaling applications based on incoming MQTT messages. Not intended to support bidirectional communications
  - **How do you want to scale:** Based on volume of messages on a particular topic
  - **Authentication:** Two options are [username/password auth](https://www.hivemq.com/blog/mqtt-security-fundamentals-authentication-username-password/) and [advanced (x.509)](https://www.hivemq.com/blog/mqtt-security-fundamentals-advanced-authentication-mechanisms/) auth. username/password auth should be sufficient for now, given that you can already place credentials in a secret and pass them into a `ScaledObject`

**Comment by arschles** [comment]
I would be happy to get started on this.

**Comment by andschneider** [comment]
I'd be interested in helping out as well!

**Comment by zroubalik** [comment]
I like the idea. It's up to you guys, who wants to work on this! :) Thanks!

**Comment by iennae** [comment]
@zroubalik what does the "needs discussion" label mean? It's ok to start implementing right?

**Comment by andschneider** [comment]
I started looking into making the scaler and was trying to figure out what to use as the metric value. 

Since MQTT brokers don't have the concept of a queue, using queue length or similar wouldn't work. I was then thinking of using number of messages received over a certain period of time. However, I'm not sure if time based scalers are possible. @arschles Is this what you meant with "volume of messages"?

**Comment by arschles** [comment]
I didn't have anything specific in mind regarding "volume of messages" -  just an abstract idea. I think that number of messages received over a period of time could work. another idea is to maintain a rolling average of messages received over a window. according to you, @andschneider , you'd need to have KEDA be a subscriber to the queue - not sure if that's ok to do. @zroubalik thoughts on that part?

**Comment by zroubalik** [comment]
> @zroubalik what does the "needs discussion" label mean? It's ok to start implementing right?

that means that this issue needs some clarification/specification/proposal on how it will be implemented, basically the stuff that you can see is happening.

**Comment by zroubalik** [comment]
@arschles so your idea is, that KEDA will subscribe to the topic and will do the calculation (messages received over a window etc) ?

**Comment by andschneider** [comment]
I'm by no means an MQTT expert either so take this all with a grain of salt, but I did a little more research and it seems that some brokers do expose metrics. As it's not part of the MQTT spec it's up to the broker's implementation for what they are though. For example, [mosquitto](https://mosquitto.org/man/mosquitto-8.html) and [HiveMQ ](https://www.hivemq.com/docs/hivemq/4.4/user-guide/monitoring.html#metrics) seem to be popular choices and they expose different metrics.

To keep the scaler as general as possible my thought was to stick to something in the spec. There's a thing called _retained messages_ which is flag set on a published message that basically tells the broker to always store one (and only one) message for a topic, until it gets deleted. Essentially it could be used like a status message and if present the scaler would spin something up. 

Obviously a little different than using some sort of messages in a time span/window approach - but it gets around having the scaler subscribed to a topic. @arschles @zroubalik what do you guys think?

**Comment by arschles** [comment]
@andschneider would you be up for getting this started? A lot of my time is taken getting the [http-add-on](https://github.com/kedacore/http-add-on) started

**Comment by andschneider** [comment]
@arschles Yep, if everyone is cool with the direction I'll have some time towards the end of this week to get it going.

**Comment by andschneider** [comment]
@arschles Apologies for the delay - I finally started working on it this weekend. 

Here's a [repo](https://github.com/andschneider/keda-mqtt-example) with an example setup for running a MQTT broker and pub/sub components. The basic scaler seems to be working (its on the `mqtt-scaler` branch of my Keda fork). Main things left to do are auth, higher QoS messages, and general robustness.

**Comment by tomkerkhove** [comment]
Great to hear, thanks!

**Comment by andschneider** [comment]
Quick update - with holidays and whatnot I haven't been able to spend much time on this. Next week is finally looking all clear though and I should be able to make some progress!

**Comment by tomkerkhove** [comment]
That's great, thanks!

**Comment by arschles** [comment]
@zroubalik it doesn't seem like there's much support behind the MQTT-based scaler. should we close this or let the stale bot do it in due time?

**Comment by zroubalik** [comment]
@arschles let's keep this open as it is a valid task, maybe someone else will pick it up in the future 🤞

**Comment by arschles** [comment]
Sounds good!

**Comment by vainkop** [comment]
MQTT scaler would be definitely good for AWS iot core processing https://docs.aws.amazon.com/iot/latest/developerguide/mqtt.html

**Comment by MrMithun101** [comment]
Hi @zroubalik @arschles - I'd like to pick this up, if it's still open and no one else is actively on it.

I've read through the earlier discussion, the sticking point seemed to be that MQTT has no native queue-length concept, so there's no obvious "point-in-time" metric the way there is for a message queue. My initial thought is to go with a time-windowed message count (subscribe to the topic, count messages received over a configurable windowSeconds) as the primary metric, since that avoids needing KEDA to hold open a long-lived subscription just to inspect broker-internal state, and it maps cleanly onto KEDA's existing target-value/HPA math. Happy to also look at the retained-message idea @andschneider raised as a secondary/activation signal if that's useful.

Before I put time into an implementation, I wanted to explicitly check in and get assigned first rather than assume, I want to make sure my approach lines up with what you'd want to see before I start building. Let me know if this is still up for grabs and if the direction above sounds reasonable.

**Comment by zroubalik** [comment]
@MrMithun101 this is still for grabs, though I am not expert in the area, so I am not sure. Assigning to you, let's see if you can come up with something reasonable.

**Comment by MrMithun101** [comment]
Quick design update before I start building, to make sure this lines up with what you'd want to see.

Since MQTT brokers don't expose queue depth the way a message queue does, this scaler needs to actively watch the topic rather than poll for a number. Concretely, I'm planning to implement it as a KEDA PushScaler (the same pattern used internally by the gRPC external scaler) rather than a plain poll-based Scaler:

Run(ctx, active chan<- bool) opens the MQTT connection and subscribes to the topic for the lifetime of the scaler, maintaining a rolling window count of messages received.
GetMetricsAndActivity simply reads the current window count when KEDA polls it — no live querying needed at that point.
The active channel lets a retained-message signal push activation immediately, rather than waiting for the next poll cycle.

Planned config fields: brokerAddress, topic, windowSeconds (default TBD, open to suggestion), queryValue (scaling threshold), plus standard TLS/auth fields matching other scalers.

Two open questions I'd appreciate your take on:

QoS level to subscribe at — QoS 1 risks duplicate counts on reconnect, QoS 0 risks silent message loss. Leaning QoS 1 with dedup awareness documented as a known limitation for v1.
Should retained-message handling ship in the first PR, or land as a fast-follow once the core windowed-count scaler is reviewed?

Will start on the implementation once I hear back, or after a couple of days if there's no strong objection.
